src/PolyType.Roslyn/ModelGenerator/TypeDataModelGenerator.Dictionary.cs

```diff
@@ -46,81 +47,95 @@ private bool TryMapDictionary(ITypeSymbol type, ImmutableArray<AssociatedTypeMod
             return false; // Not a dictionary type
         }
 
-        if (namedType.Constructors.FirstOrDefault(ctor => ctor.Parameters.Length == 0 && !ctor.IsStatic && IsAccessibleSymbol(ctor)) is { } ctor &&
+        // .ctor()
+        if (namedType.Constructors.FirstOrDefault(ctor => ctor is { Parameters: [], IsStatic: false } && IsAccessibleSymbol(ctor)) is { } ctor &&
             ContainsSettableIndexer(type, keyType, valueType))
         {
             constructionStrategy = CollectionModelConstructionStrategy.Mutable;
             factoryMethod = ctor;
+
+            // .ctor(I[Equality]Comparer<TKey>)
+            factoryMethodWithComparer = namedType.Constructors.FirstOrDefault(ctor
+                => ctor is { Parameters: [IParameterSymbol only] }
+                && IsAccessibleSymbol(ctor)
+                && ClassifyConstructorParameter(only, keyType, valueType) is CollectionConstructorParameterType.Comparer or CollectionConstructorParameterType.EqualityComparer);
         }
-        else if (namedType.Constructors.FirstOrDefault(ctor =>
-            IsAccessibleSymbol(ctor) &&
-            ctor.Parameters is [{ Type: INamedTypeSymbol parameterType }] &&
-            SymbolEqualityComparer.Default.Equals(parameterType.ConstructedFrom, KnownSymbols.ReadOnlySpanOfT) &&
-            parameterType.TypeArguments is [INamedTypeSymbol elementType] &&
-            SymbolEqualityComparer.Default.Equals(elementType.ConstructedFrom, KnownSymbols.KeyValuePairOfKV) &&
-            elementType.TypeArguments is [INamedTypeSymbol k, INamedTypeSymbol v] &&
-            SymbolEqualityComparer.Default.Equals(k, keyType) &&
-            SymbolEqualityComparer.Default.Equals(v, valueType)) is IMethodSymbol ctor2)
+
+        // .ctor(ReadOnlySpan<KeyValuePair<K, V>>)
+        if (factoryMethod is null &&
+            namedType.Constructors.FirstOrDefault(ctor
+                => IsAccessibleSymbol(ctor)
+                && ctor is { Parameters: [{ Type: { } parameterType }] }
+                && IsSpanOfKeyValuePair(parameterType, keyType, valueType)) is IMethodSymbol ctor2)
         {
             constructionStrategy = CollectionModelConstructionStrategy.Span;
             factoryMethod = ctor2;
+
+            // .ctor(ReadOnlySpan<KeyValuePair<K, V>>, I[Equality]Comparer<T>) or .ctor(I[Equality]Comparer<T>, ReadOnlySpan<KeyValuePair<K, V>>)
```

Comment:
Unless I've already done this, we should add an issue tracking support for the final shape of dictionary collection builders: https://github.com/dotnet/runtime/issues/114090

Reply:
You already have one: https://github.com/eiriktsarpalis/PolyType/issues/148